include/triton/core/tritonbackend.h

```diff
@@ -1359,6 +1360,86 @@ TRITONBACKEND_ISPEC TRITONSERVER_Error* TRITONBACKEND_ModelInstanceExecute(
+///
+/// \param backend The backend.
+/// \param model The model. 'nullptr' may be provided if no model is associated.
+/// \param type The attribute to query about.
+/// \param in_params The array of input parameters as a complement of the
+/// attribute inquiry.
+/// \param in_count The number of input parameters.
+/// \param backend_attributes The backend attribute. See
+/// TRITONBACKEND_BackendAttribute on how it should be set for different 'type'.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_ISPEC TRITONSERVER_Error* TRITONBACKEND_GetBackendAttribute(
+    TRITONBACKEND_Backend* backend, TRITONBACKEND_Model* model,
+    const TRITONBACKEND_AttributeType type,
+    const TRITONSERVER_Parameter** in_params, const uint64_t in_count,
+    TRITONBACKEND_BackendAttribute** backend_attributes);
+
+/// TRITONBACKEND_BackendAttribute
+///
+/// API to create, modify, or retrieve attributes associated with a backend.
+///
+
+/// Create a new backend attributes object. The caller takes ownership of
+/// the TRITONBACKEND_BackendAttribute object and must call
+/// TRITONBACKEND_BackendAttributeDelete to release the object.
+///
+/// \param backend_attributes Returns the new backend attributes object.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONSERVER_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_BackendAttributeNew(
+    TRITONBACKEND_BackendAttribute** backend_attributes);
+
+/// Delete a backend attributes object.
+///
+/// \param backend_attributes The backend_attributes object.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONSERVER_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_BackendAttributeDelete(
+    TRITONBACKEND_BackendAttribute* backend_attributes);
+
+///
+/// TRITONBACKEND_ATTRIBUTE_PREFERRED_INSTANCE_GROUP
+///
+/// Set the preferred instance group of the backend. This function should be
+/// called for TRITONBACKEND_ATTRIBUTE_PREFERRED_INSTANCE_GROUP. This function
+/// can be called multiple times to cover different instance group kinds that
+/// the backend supports, given the priority order that the first call describes
+/// the most preferred group. In the case where instance group are not
+/// explicitly provided, Triton will use this attribute to create model
+/// deployment that aligns more with the backend preference.
+///
+/// \param backend_attributes The backend attributes object.
+/// \param kind The kind of the instance group.
+/// \param count The number of instances per device. Triton default will be used
+/// if 0 is provided.
+/// \param device_ids The devices where instances should be available. Triton
+/// default will be used if 'nullptr' is provided.
+/// \param id_count The number of devices in 'device_ids'.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONSERVER_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_BackendAttributeSetPreferredInstanceGroup(
```

Comment:
I was wondering maybe this should be a model attribute instead of a backend attribute? For example, a backend may decide that the best instance group for this model is CPU and the best instance group for another model is GPU?

Another complete different way of approaching this problem would be to allow setting instance_group settings in the auto-complete API. This way we wouldn't need a new API and the backend can indicate the instance group preference as a part of auto-completing the model configuration.

Reply:
The decision on not modifying current auto-complete API behavior is that new API will be more clear on what is changed. The other thing that I was thinking is that if it is done within auto-complete API, the backend needs to check available devices to make the correct decision instead of setting a set of preferred settings and let Triton core to match them.